pinot-core/src/main/java/org/apache/pinot/core/io/writer/impl/OffHeapByteArrayStore.java

```diff
@@ -0,0 +1,205 @@
+ * an index. No verification is made as to whether the value added already exists or not.
+ * Empty byte arrays are supported.
+ *
+ * @note The class is thread-safe for single writer and multiple readers.
+ *
+ * Byte arrays (values) are stored as below:
+ *
+ *                  __________________________________
+ *                  |  start offset of array  1      |
+ *                  |  start offset of array  2      |
+ *                  |        .....                   |
+ *                  |  start offset of array  N      |
+ *                  |                                |
+ *                  |         UNUSED                 |
+ *                  |                                |
+ *                  |  Array N .....                 |
+ *                  |          .....                 |
+ *                  |          .....                 |
+ *                  |  Array N-1                     |
+ *                  |          .....                 |
+ *                  |          .....                 |
+ *                  |  Array 0 .....                 |
+ *                  |          .....                 |
+ *                  |          .....                 |
+ *                  |________________________________|
+ *
+ *
+ * We fill the buffer as follows:
+ * - The values are added from the bottom, each new value appearing nearer to the top of the buffer, leaving no
+ *   room between them. Each value is stored as a sequence of bytes.
+ *
+ * - The start offsets of the byte arrays are added from the top. Each start offset is stored as an integer, taking 4 bytes.
+ *
+ * Each time we want to add a new value, we check if we have space to add the length of the value, and the value
+ * itself. If we do, then we compute the start offset of the new value as:
+ *
+ *    new-start-offset = (start offset of prev value added) - (length of this value)
+ *
+ * The new start offset value is stored in the offset
+ *
+ *    buffer[numValuesSoFar * 4]
+ *
+ * and the value itself is stored starting at new-start-offset
+ *
+ */
+public class OffHeapByteArrayStore implements Closeable {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(OffHeapByteArrayStore.class);
+
+  private final PinotDataBuffer _pinotDataBuffer;
+  private final ByteBuffer _byteBuffer;
+  private final long _size;
+
+  private int _numValues = 0;
+  private int _availEndOffset;  // Exclusive
+
+  /**
+   * This constructor is useful when the values are not known ahead of time. If the caller uses this
+   * constructor, then it is expected that the caller invokes {@link #add(byte[])} method until it returns -1
+   * at which point it is known that no more strings can be added to the buffer.
+   *
+   * @param size
+   * @param memoryManager
+   * @param allocationContext
+   */
+  public OffHeapByteArrayStore(long size, PinotDataBufferMemoryManager memoryManager, String allocationContext) {
+    if (size >= Integer.MAX_VALUE) {
```

Comment:
@mcvsubbu Though remote case, should we fail to create the buffer rather than silently trimming the size to `MAX_VALUE - 1`? Maybe making the size itself an `int` might make this explicit.